### ISSUE TITLE
Issue in WebRTC streaming due to race in dispatching the task in Mainthread

### ISSUE THREAD

**Issue by kkanag314** [Metrological/WebKitForWayland#805]
We are observing a issue in Webrtc video streaming when both audio and video is enabled in remote peer.
The libwebrtc implementation has logic to post a task to add the tracks in mediastream in the MainThread.

 void LibWebRTCMediaEndpoint::OnTrack(rtc::scoped_refptr<webrtc::RtpTransceiverInterface> transceiver)
 {
     if (!RuntimeEnabledFeatures::sharedFeatures().webRTCUnifiedPlanEnabled())
         return;
     callOnMainThread([protectedThis = makeRef(*this), transceiver = WTFMove(transceiver)]() mutable {
         if (protectedThis->isStopped())
             return;
         protectedThis->newTransceiver(WTFMove(transceiver));
     });
 }

Since the order of the tracks from remote peer is Audio,Video,Data,the task to add the audiotrack in mediastream is posted followed by Video,data

Once the Dispatcher timer is fired,The task to add the audiotrack in mediastream(which is scheduled first) is being dispatched. Once the audio track is added in the stream,Track event is fired and application configures the srcobject attribute.
The srcobject is configured in HTMLMediaElement and selectMediaresource() has been called.
In Resource Selection algorithm,the task has been scheduled in a genericQueue. 

It seems that the MainThread's event source & GenericTaskQueue's event sources are of same priority.The priority for the Main Thread and generic taskqueue is  configured based on the RunLoopSourcePriority::MainThreadDispatcherTimer (-60).Due to this, there is races in dispatching Mainthread function queue and GenericTaskQueue.

In Mainthread,the maxRunLoopSuspensionTime is set as 50ms. In issue cases, Before adding the video track in the media stream(which is scheduled in the Mainthread fucntionqueue),Dispatching is suspended because it exceeds the maxRunLoopSuspensionTime. Due to the same priority of MainThread's event source & GenericTaskQueue's event sources, the task posted in the GenericTaskQueue(m_resourceSelectionTaskQueue) is dispatched which lead to the pipeline construction. Since the mediastream only has audiotrack,only one appsrc is generated for the mediastream and corresponding elements for audio playback is added in the pipeline and moved to PLAYING state. Once this task is completed,the task scheduled in  MainThread's event source is started dispatching and videotrack is added in the mediastream.Since the pipeline is already constrcuted,Due to the unavailability of appsrc and other gstelement for the videotrack,Only black screen is observed.

**Comment by kkanag314** [comment]
@woutermeek Could you please help to review this issue. If any additional information's are needed, please let me know

**Comment by goruklu** [comment]
we also see similar issue with some the streaming apps. @woutermeek  can you please ask your team to take a look at this ? We can help with testing if needed.

**Comment by woutermeek** [comment]
Sure thing, we will look into it soon.

**Comment by magomez** [comment]
I think the problem here is that limit of 50ms when dispatching MainThread functions. I'm afraid that value is an heuristic that's valid for most of hardware devices, but it might be too short for low end devices where processing the callbacks is slower, and where UI delays of more than 0.1 seconds are acceptable (they're the norm actually). Due to this, in some cases not all the enqueued functions may be called, causing this behavior. I would be happy to increase that limit, but it would be awesome if you could do some experiments on your side to find the smallest value that fixes your problem (to avoid the side effects of values too big). I think we can safely go to 100ms or maybe 200ms, but that requires some testing.

Just for reference, the var to modify is maxRunLoopSuspensionTime inside Source/WTF/wtf/MainThread.cpp.

Please, @kkanag314 and @goruklu , give it some testing and tell me which value makes sense to you. Thanks!

**Comment by kkanag314** [comment]
OK @magomez.  We will validate with increased maxRunLoopSuspensionTime value and share our observation

**Comment by kkanag314** [comment]
@magomez  We have done the validation with maxRunLoopSuspensionTime as 150ms.The issue is not observed with this value.We have done smoke testing for premium webapps in highend /lowend devices. We don't observe any issues so far.

**Comment by magomez** [comment]
I've just pushed 47c3227f301b09a49707a71b9de2061d2709aa9e to the wpe-2.22 branch with that change. I was going to add some env var to get the value, but I think that this change should not cause a noticeable difference in most of the platforms, so I've just set the 150ms. If some other users find regressions with this, I'll replace the hardcoded value with the environment variable so it can be set per platform.
